scripts/Muon/GUI/Common/grouping_table_widget/grouping_table_widget_presenter.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import (absolute_import, division, print_function)
+
+import re
+
+from Muon.GUI.Common.utilities import run_string_utils as run_utils
+from Muon.GUI.Common.muon_group import MuonGroup
+
+
+class GroupingTablePresenter(object):
+
+    def __init__(self, view, model):
+        self._view = view
+        self._model = model
+
+        self._view.on_add_group_button_clicked(self.handle_add_group_button_clicked)
+        self._view.on_remove_group_button_clicked(self.handle_remove_group_button_clicked)
+
+        self._view.on_user_changes_group_name(self.validate_group_name)
+        self._view.on_user_changes_detector_IDs(self.validate_detector_ids)
+
+        self._view.on_table_data_changed(self.handle_data_change)
+
+        self._dataChangedNotifier = lambda: 0
+
+    def show(self):
+        self._view.show()
+
+    def on_data_changed(self, notifier):
+        self._dataChangedNotifier = notifier
+
+    def notify_data_changed(self):
+        self._dataChangedNotifier()
+
+    def _is_edited_name_duplicated(self, new_name):
+        is_name_column_being_edited = self._view.grouping_table.currentColumn() == 0
+        is_name_unique = (sum(
+            [new_name == name for name in self._model.group_and_pair_names]) == 0)
+        return is_name_column_being_edited and not is_name_unique
+
+    def validate_group_name(self, text):
+        if self._is_edited_name_duplicated(text):
+            self._view.warning_popup("Groups and pairs must have unique names")
+            return False
+        if not re.match(run_utils.valid_name_regex, text):
+            self._view.warning_popup("Group names should only contain digits, characters and _")
+            return False
+        return True
+
+    def validate_detector_ids(self, text):
+        if re.match(run_utils.run_string_regex, text):
+            return True
+        self._view.warning_popup("Invalid detector list.")
+        return False
+
+    def disable_editing(self):
+        self._view.disable_editing()
+
+    def enable_editing(self):
+        self._view.enable_editing()
+
+    def add_group(self, group):
+        """Adds a group to the model and view"""
+        if self._view.num_rows() > 19:
```

Comment:
can we make this 19 and the 20 below use some sort of constant that we define at the top of the file?